lib/Parse/ParseExpr.cpp

```diff
@@ -3255,18 +3255,14 @@ ExprResult Parser::ParseBoundsExpression() {
   return Result;
 }
 
-// Parse a generic type argument list.  The suffix of postfix expression can
-// have the form '<' type name 1, ... type name n '>', in which case it is a
-// generic type argument list.
-//
-// For now, handle only the case where Res is a generic DeclRef.
-//
-// TODO: We need to handle the case where Res is an expression with
-// generic type.  However, we don't have the AST support for this yet.
-//
-// Returns false if parsing succeeded and true if an error occurred.
-ExprResult Parser::ParseGenericTypeArgumentList(ExprResult Res, SourceLocation Loc) {
-  SmallVector<DeclRefExpr::GenericInstInfo::TypeArgument, 4> typeArgumentInfos;
+// Parse a generic type argument list of the form type_name_1, ..., type_name_n '>'.
+// Notice the initial '<' must have already been parsed.
+// This is re-used for both generic functions and generic structs.
+// Return false if parsing succeeds, in which case the 'typeArgs' is also populated.
+// Return true if parsing fails.
+std::pair<bool, Parser::TypeArgVector> Parser::ParseGenericTypeArgumentList(SourceLocation Loc) {
+  Parser::TypeArgVector typeArgumentInfos;
+  auto err = std::make_pair<>(true, Parser::TypeArgVector());
   bool firstTypeArgument = true;
```

Comment:
bool firstTypeArgument: This can also be declared as auto.

Reply:
We should see what the clang style guide says on this one.  auto is just as long to type as  bool :).

Reply:
auto it is